### PR TITLE
DOC: Correct subplot() doc

### LINKED ISSUE #7393: `subplot()`: incorrect description of deletion of overlapping axes in the docs
The docstring for `plt.subplot` states that
```
       Creating a new subplot with a position which is entirely inside a
       pre-existing axes will trigger the larger axes to be deleted::

          import matplotlib.pyplot as plt
          # plot a line, implicitly creating a subplot(111)
          plt.plot([1,2,3])
          # now create a subplot which represents the top plot of a grid
          # with 2 rows and 1 column. Since this subplot will overlap the
          # first, the plot (and its axes) previously created, will be removed
          plt.subplot(211)
          plt.plot(range(12))
          plt.subplot(212, facecolor='y') # creates 2nd subplot with yellow background

       If you do not want this behavior, use the
       :meth:`~matplotlib.figure.Figure.add_subplot` method or the
       :func:`~matplotlib.pyplot.axes` function instead.
```
(which I think is a rather silly behavior but whatever -- seems inherited from MATLAB, in fact).

The actual behavior (which is also MATLAB's behavior) is to delete any axes that *overlaps* the newly added axes, which can be easily checked from the code
```
    for other in fig.axes:
        if other==a: continue
        if bbox.fully_overlaps(other.bbox): # <-- here
            byebye.append(other)
    for ax in byebye: delaxes(ax)
```
and also by running e.g. `plt.subplot(221); plt.subplot(335)` (the first subplot does get deleted).

Making this a doc issue rather than requesting for the behavior to be changed because I don't think either behavior really makes sense anyways :-)

Noticed while investigating #6096 -- this is the mechanism that leads to one of the axes being deleted (note, though, that regardless of this, there still(?) seems to be an issue with incremental tight-layouting in #6096).

### PR DESCRIPTION
**Pull request by petehuang**
Fixes #7393

### CODE REVIEW

File: lib/matplotlib/pyplot.py [hunk 1]
```diff
@@ -976,8 +976,8 @@ def subplot(*args, **kwargs):
 
     .. note::
 
-       Creating a new subplot with a position which is entirely inside a
-       pre-existing axes will trigger the larger axes to be deleted::
+       Creating a new subplot with axes that overlap pre-existing axes will
+       trigger the pre-existing axes to be deleted::
```
Comment (anntzer) [inline lib/matplotlib/pyplot.py]: Something like "Creating a subplot will delete any pre-existing subplot that overlaps with it." seems clearer (possibly add "(if any)" at the end).

**Comment by petehuang** [comment]
Good suggestion @anntzer

**Comment by anntzer** [comment]
lgtm

File: lib/matplotlib/pyplot.py [hunk 2]
```diff
@@ -976,8 +976,8 @@ def subplot(*args, **kwargs):
 
     .. note::
 
-       Creating a new subplot with a position which is entirely inside a
-       pre-existing axes will trigger the larger axes to be deleted::
+       Creating a subplot will delete any pre-existing subplot that overlaps
```
Comment (tacaswell) [inline lib/matplotlib/pyplot.py]: Overlaps or fully overlaps?

**Comment by NelleV** [comment]
Thanks @petehuang

**Comment by efiring** [comment]
I challenge anyone reading the proposed modification of the docstring, without poring over the code, to explain what "fully overlaps" means.  Before reading the code, I was imagining that it might mean "encloses", as the original form of the docstring implied. Based on the Bbox method code, however, it means that the interiors intersect--that is, the intersection of the two rectangles encloses a finite area.  Hence the subplot code is deleting any pre-existing subplots that have a finite intersection with the new subplot. Coinciding edges are not enough, because then the intersection encloses no area.

**Comment by petehuang** [comment]
Ok let me paraphrase to make sure I understand correctly: you're saying that it's not determined by axes lying over one another, but it's that the rectangles that the axes create intersect in some way?

I just want to make sure we get this explanation and the docstring in plain english

**Comment by efiring** [comment]
The practical problem is coming from the adverb "fully".  The "overlaps" part is fine.  The subtle distinction is that a shared boundary does not count as an "overlap".  You could change "that fully overlaps with it" to "that has a finite area of overlap with it".  Or "that overlaps with it beyond sharing a boundary".

**Comment by NelleV** [comment]
I knew what the function did (and I find this behavior extremely strange…).
I would actually raise a warning to the user when that happens, as this behavior is quite unexpected.

**Comment by efiring** [comment]
Thank you!

**Comment by QuLogic** [comment]
Backported to `v2.0.x` as feeba1f6eacd0dd4bee1ba3932143a52f24af2ad.
